src/data_chain/vote_accumulator.rs

```diff
@@ -39,33 +38,136 @@ pub enum AccumulationReturn<T> {
+
+        match self.blocks.entry(vote.payload().clone()) {
+            Entry::Vacant(entry) => {
+                let _ = entry.insert(State {
+                    block: Block::new(vote, node_info)?,
+                    timestamp: Instant::now(),
+                });
+            }
+            Entry::Occupied(mut entry) => {
+                let valid_before = entry
+                    .get()
+                    .block
+                    .get_block_state(valid_nodes)
+                    .valid_or_full();
+                let valid_after = entry
+                    .get_mut()
+                    .block
+                    .add_vote(vote, node_info, valid_nodes)?
+                    .valid_or_full();
+
+                if valid_after {
+                    if valid_before {
+                        let proof = vote.proof(node_info)?;
+                        results.push(AccumulationReturn::ValidProof(proof));
+                    } else {
+                        let block = entry.get().block.clone();
+                        results.push(AccumulationReturn::ValidBlock(block));
+                    }
+                }
+            }
+        }
+
+        Ok(results)
     }
 
     // FIXME - remove these
     #[allow(unused)]
-    #[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
     pub fn add_block(
         &mut self,
-        _block: Block<T>,
+        block: Block<T>,
+        valid_nodes: &BTreeSet<PublicInfo>,
     ) -> Result<Vec<AccumulationReturn<T>>, RoutingError> {
-        Ok(vec![])
+        let mut results = self.expire(valid_nodes);
+
+        match self.blocks.entry(block.payload().clone()) {
+            Entry::Vacant(entry) => {
+                if block.get_block_state(valid_nodes).valid_or_full() {
+                    results.push(AccumulationReturn::ValidBlock(block.clone()));
+                }
+
+                let _ = entry.insert(State {
+                    block,
+                    timestamp: Instant::now(),
+                });
+            }
+            Entry::Occupied(mut entry) => {
+                let valid_before = entry
+                    .get()
+                    .block
+                    .get_block_state(valid_nodes)
+                    .valid_or_full();
+                let mut valid_after = false;
+                let mut new_proofs = Vec::new();
+
+                for proof in block.proofs() {
+                    match entry.get_mut().block.add_proof(*proof, valid_nodes) {
+                        Ok(state) => {
+                            valid_after = state.valid_or_full();
+                            new_proofs.push(*proof);
+                        }
+                        /// Duplicate votes are ignored.
+                        Err(RoutingError::DuplicateSignatures) => (),
+                        Err(error) => return Err(error),
+                    }
+                }
+
+                if valid_after {
```

Comment:
can we do this:
```
for proof in block.proofs() {
    let _ =  entry.get_mut().block.add_proof(*proof, valid_nodes);
}
if entry.get().block.get_block_state(valid_nodes).valid_or_full() {
   if valid_before ...
}
```

Reply:
I still need to check the result of `add_proof` for errors and also to collect the new proofs, so I may as well use it in the `if` too. This also avoids the `get_block_state` call, making this approach slightly more efficient. I'd prefer to do it the way I'm doing it unless you think it's actually incorrect.

Reply:
the `new_proofs` only needs to be collected when `valid_before` is true, and you don't need to carry out `state.valid_or_full()` each time adding proof.
Regarding the `error` check, will it make more sense to `restore to the previous state` or `ignore then to continue`, instead of the current `partially updated then break` ?

So, can we do something like:
```
for proof in block.proofs() {
    if valid_before && entry.get_mut().block.add_proof(*proof, valid_nodes).is_ok() {
        results.push(AccumulationReturn::ValidProof(*proof));
    }
}
if !valid_before && entry.get().block.get_block_state(valid_nodes).valid_or_full() {
    results.push(AccumulationReturn::ValidBlock(entry.get().block.clone()));
}
```